### PR TITLE
[WIP] TClass::GetClass: no interpreter lookups for fundamental types

### LINKED ISSUE #9029: Repeated calls to TClass::GetClass/TClass::ResolveTypedef hoarding memory
Hi everyone,

We came across this issue in ATLAS recently. There were similar reports in the past, such as [ROOT-7254](https://sft.its.cern.ch/jira/browse/ROOT-7254), but there is none that's still open as far as I could see. 

### Describe the bug
Repeatedly calling `TClass::GetClass`/`TClass::ResolveTypedef` w/ an input such as `std::vector<int>::value_size` seems to create memory hoarding.

### To Reproduce
Either of these python scripts should reproduce the issue:

```
import ROOT

while True:
    ROOT.TClass.GetClass("std::vector<int>::value_size")
```

```
import ROOT

while True:
    ROOT.TClassEdit.ResolveTypedef("std::vector<int>::value_size")
```

### Setup
Tested with both ROOT versions 6.24/00 and 6.24/06 in the ATLAS production environment.

Please let us know, many thanks.

Best,
Serhan

### PR DESCRIPTION
**Pull request by dpiparo**
Add yet another fence in TClass::GetClass to avoid lookups and memory consumption. This PR aims to fix #9029. Give the sophisticated implementation of TClass::GetClass, perhaps it would be good to collect some feedback, especially by @pcanal .

# This Pull request:
This PR avoids lookups and parsing in some cases.
One of the principles of the TClass::GetClass method implementation is to avoid as much as possible. 

## Changes or fixes:
This commit adds yet another fence in TClass::GetClass, checking if the name in input is the name of a known fundamental type or typedef to it.
In order to avoid code duplication, a routine previously available within the implementation of TClassEdit has been made available with a public API.

## Checklist:

- [v ] tested changes locally
- [v ] updated the docs (if necessary)

This PR fixes #9029

### CODE REVIEW

File: core/meta/src/TClass.cxx [hunk 1]
```diff
@@ -3046,6 +3046,20 @@ TClass *TClass::GetClass(const char *name, Bool_t load, Bool_t silent, size_t hi
       // continue as before ...
    }
 
+   // Check if this is a fundamental type already in the typesystem to delay
+   // as much as possible any lookup
+   if (!cl) {
+      std::string name_no_std = name;
+      TClassEdit::RemoveStd(name_no_std);
+      auto lot = gROOT->GetListOfTypes();
+      if(lot->Contains(name_no_std.c_str()))
```
[the remark this answers is not in the document]
Comment (dpiparo) [reply core/meta/src/TClass.cxx:3055]: thanks for the feedback! I am proposing a new version of the code, also considering the failing tests.

**Comment by dpiparo** [comment]
Hello @phsft-bot build just on ROOT-fedora37, ROOT-fedora38, ROOT-ubuntu2204

File: core/meta/src/TClass.cxx [hunk 2]
```diff
@@ -3047,13 +3047,35 @@ TClass *TClass::GetClass(const char *name, Bool_t load, Bool_t silent, size_t hi
...
 
+   // Check if this is a fundamental type already in the typesystem to delay
+   // as much as possible any lookup
    if (!cl) {
       {
          TInterpreter::SuspendAutoLoadingRAII autoloadOff(gInterpreter);
          TClassEdit::GetNormalizedName(normalizedName, name);
       }
+      std::string name_no_std = name;
+      TClassEdit::RemoveStd(name_no_std);
```
Comment (pcanal) [inline core/meta/src/TClass.cxx:3059]: What is the advantage of using this `name_no_std` rather than the `normalizedName`?

**Comment by dpiparo** [comment]
@phsft-bot build

**Comment by dpiparo** [comment]
@phsft-bot build , still too many missing containers...

**Comment by dpiparo** [comment]
Thanks for all the feedback. A more comprehensive solution to avoid spurious lookups triggered by TClass::GetClass will be proposed.

**Pull request closed without merging**